DependencyInjection/Compiler/CacheCompatibilityPass.php

```diff
@@ -0,0 +1,100 @@
+
+    public function process(ContainerBuilder $container): void
+    {
+        foreach (array_keys($container->findTaggedServiceIds(DoctrineMongoDBExtension::CONFIGURATION_TAG)) as $id) {
+            /** @var array{0: string, 1: mixed[]} $methodCall */
+            foreach ($container->getDefinition($id)->getMethodCalls() as $methodCall) {
+                $methodName = $methodCall[0];
+
+                if (! isset(self::CACHE_SETTER_METHODS_PSR6_SUPPORT[$methodName])) {
+                    continue;
+                }
+
+                $methodArgs   = $methodCall[1];
+                $definitionId = (string) $methodArgs[0];
+                $aliasId      = null;
+                if ($container->hasAlias($definitionId)) {
+                    $aliasId      = $definitionId;
+                    $definitionId = (string) $container->getAlias($aliasId);
+                }
+
+                $shouldBePsr6 = self::CACHE_SETTER_METHODS_PSR6_SUPPORT[$methodName];
+
+                $this->wrapIfNecessary($container, $aliasId, $definitionId, $shouldBePsr6);
+            }
+        }
+    }
+
+    private function createCompatibilityLayerDefinition(ContainerBuilder $container, string $definitionId, bool $shouldBePsr6): ?Definition
+    {
+        $definition = $container->getDefinition($definitionId);
+
+        while (! $definition->getClass() && $definition instanceof ChildDefinition) {
+            $definition = $container->findDefinition($definition->getParent());
+        }
+
+        if ($shouldBePsr6 === is_a($definition->getClass(), CacheItemPoolInterface::class, true)) {
+            return null;
+        }
+
+        $targetClass   = CacheProvider::class;
+        $targetFactory = DoctrineProvider::class;
+
+        if ($shouldBePsr6) {
+            $targetClass   = CacheItemPoolInterface::class;
+            $targetFactory = CacheAdapter::class;
+
+            trigger_deprecation(
+                'doctrine/mongodb-odm-bundle',
+                '4.4',
+                'Configuring doctrine/cache is deprecated. Please update the cache service "%s" to use a PSR-6 cache.',
+                $definitionId
+            );
+        }
+
+        return (new Definition($targetClass))
+            ->setFactory([$targetFactory, 'wrap'])
+            ->addArgument(new Reference($definitionId));
+    }
+
+    private function wrapIfNecessary(ContainerBuilder $container, ?string $aliasId, string $definitionId, bool $shouldBePsr6): void
+    {
+        $compatibilityLayer = $this->createCompatibilityLayerDefinition($container, $definitionId, $shouldBePsr6);
+
+        if ($compatibilityLayer === null) {
+            return;
+        }
+
+        $aliasId = $aliasId ?? $definitionId;
+
+        $compatibilityLayerId = $definitionId . '.compatibility_layer';
+        $container->setAlias($compatibilityLayerId, $aliasId);
+
+        $container->setDefinition($compatibilityLayerId, $compatibilityLayer);
```

Comment:
Not sure about this, but if I had understood correctly, if `$compatibilityLayer` is not `null`, it is a definition implementing PSR-6 to replace the current definition, so should we replace the current alias with the compatibilityLayer one (flipping arguments)?:

```php
$container->setAlias($aliasId, $compatibilityLayerId);
```